File: sdk/js/src/transports/transport.ts
/**
 * GRAFT Transport Interface
 *
 * All transport implementations must satisfy this interface.
 * The protocol layer calls these methods without knowing
 * whether bytes travel over serial, BLE, MQTT, or WebSocket.
 */

export interface GraftTransport {
  /** Establish the connection. Resolves when ready to send/receive. */
  connect(): Promise<void>

  /** Cleanly close the connection. */
  disconnect(): Promise<void>

  /** Send a raw packet (already wire-encoded, optionally COBS-encoded). */
  send(packet: Uint8Array): Promise<void>

  /** Register a handler for incoming packets. */
  onReceive(handler: (packet: Uint8Array) => void): void

  /** Whether the transport is currently connected. */
  readonly connected: boolean

  /**
   * Whether this transport requires COBS framing.
   * Serial/BLE = true (byte stream needs delimiters)
   * MQTT/TCP/WebSocket = false (message-oriented)
   */
  readonly needsCOBS: boolean
}

File: sdk/go/transports/transport.go
// Package transports defines the Transport interface for GRAFT connections.
package transports

// Transport is the interface all GRAFT transports must implement.
type Transport interface {
	// Connect establishes the connection.
	Connect() error

	// Close cleanly closes the connection.
	Close() error

	// Send sends a raw wire-encoded packet.
	Send(packet []byte) error

	// Recv blocks until a packet is received or context is cancelled.
	Recv() ([]byte, error)

	// Connected returns whether the transport is currently connected.
	Connected() bool

	// NeedsCOBS returns whether this transport requires COBS framing.
	NeedsCOBS() bool
}
